RFC-0012-thrift-serialization-for-task-updates.md

```diff
@@ -0,0 +1,307 @@
+- If we want to change to a different binary serde, we will have to redo the process.
+
+
+
+### Option 2:  Pluggable Serde for Polymorphic Types
+
+```java
+import java.util.HashMap;
+
+public class Split
+{
+    private final ConnectorId connectorId;
+    private final ConnectorSplit connectorSplit;
+    // ... other fields and methods
+
+    public Split(...);
+}
+
+// In presto-spi, an interface for split serde 
+public interface ConnectorSplitSerializer<T extends ConnectorSplit>
+{
+    String getType();
+
+    byte[] serialize(T split);
+
+    T deserialize(byte[] data);
+}
+
+// A json serializer for hive split 
+public class HiveSplitJsonSerializer
+        implements ConnectorSplitSerializer<HiveSplit>
+{
+    private final ObjectMapper mapper;
+
+    @Override
+    public String getType()
+    {
+        return "json";
+    }
+
+    @Override
+    public byte[] serialize(HiveSplit split)
+    {
+        return mapper.writeValueAsBytes(new HiveSplitSerializable(split));
+    }
+
+    @Override
+    public HiveSplit deserialize(byte[] data)
+    {
+        HiveSplitSerializable serializable = mapper.readValue(data, HiveSplitSerializable.class);
+        return serializable.toHiveSplit();
+    }
+}
+
+//  A thrift serializer for hive split
+public class HiveSplitThriftSerializer
+        implements ConnectorSplitSerializer<HiveSplit>
+{
+    private final ThriftCodec<ThriftHiveSplit> codec;
+
+    @Override
+    public String getType()
+    {
+        return "thrift";
+    }
+
+    @Override
+    public byte[] serialize(HiveSplit split)
+    {
+        ThriftHiveSplit thriftSplit = new ThriftHiveSplit();
+        // ... populate fields ...
+        return codec.serialize(thriftSplit);
+    }
+
+    @Override
+    public HiveSplit deserialize(byte[] data)
+    {
+        ThriftHiveSplit thriftSplit = codec.deserialize(data);
+        return new HiveSplit(/* construct from thrift object */);
+    }
+}
+
+// A registry to hold the serde methods based on connector id and serde type
+public class ConnectorSplitSerializerRegistry
+{
+
+    private final Map<String, Map<String, SplitSerializer>> serializers = new HashMap();
+    
+    public void registerSerializer(Stirng connectorId, String format, SplitSerializer serializer) {...}
+    public SplitSerializer getSerializer(String connectorId, String jsonOrThrift) {...}
+}
+
+// Register the correct serde method within the corresponding connector factory
+public class HiveConnectorFactory implements ConnectorFactory {
+    @Override
+    public Connector create(...) {
+        // Register serializers
+        serializerRegistry.registerSerializer("hive", "json", new HiveSplitJsonSerializer());
+        serializerRegistry.registerSerializer("hive", "thrift", new HiveSplitThriftSerializer());
+        // ... create connector ...
+    }
+}
+
+
+// Usage
+public final class Split
+{
+    private final ConnectorSplit connectorSplit;
+    private final ConnectorSplitSerializerRegistry registry;
```

Comment:
We should probably avoid storing a reference to a serializer in every split. May be worth to simply use a registry directly